bgp: Announce LoadBalancer services via BGP Control Plane

dylandreimerink opened the pull request:
This PR builds on the initial modularization done in #22183 to provide a better path for wiring the additional resources needed for the reconciliation of services.

This PR changes pure-function reconciles into cells which adhere to a interface. This allows the new reconcile to request the service resource (via the DiffStore) without forcing other reconciles to also take these via their function parameters. Something that I plan to bring to other reconciles in the future.

Since the amount of services in clusters we have seen in the wild can exceed 750k, we have to engineer for scale. Because of this scaling issue, performing full reconciliation of the service related config is unrealistic. To curb the potential scaling issue, this PR introduces the concept of a DiffStore which wraps the `resource.Store` provided to use via resources and records which objects have changed or have been deleted since the last time we looked at this store. This allows us to apply partial updates within the existing full-reconciliation architecture for changes to service objects. The DiffStore works together with the Signaler and will trigger a reconciliation signal for services changes, but only after the initial synchronization to prevent unnecessary reconciliation stress during startup. 

```release-note
Added capability to announce LoadBalancer services via BGP Control Plane
```

dylandreimerink committed:
pkg/node: Added annotations to `types.Node`

As pref for using the local node store in BGP CP, this commit adds
the `Annotations` field to `types.Node` which is required by BGP CP.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
BGP: Made nodespecer into a cell, implemented via node local store

This commit makes `nodespecer` into a hive cell. It deduplicates the
logic to transform node resources based on the IPAM mode. This logic
is already implemented in the local node store which we now use.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bgp: Replace BGPPolicy indexers with Resources

This commit replaces the indexers with resources. Using resources has
a few advantages over the classic indexers. Namely easier sharing of
stores across the whole of cilium and a retry mechanism.

We are not yet able to profit from these advantages due to the structure
of the remaining code. Future commits can restructure the code so we
can do retries.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bgp: Config reconcilers as cells

Switched config reconcilers to be cells. The main advantage of cells is
that it allows individual reconcilers to consume additional resources
like services and endpoint information without having to modify the
function signature for other reconcilers.

The reconcilers are now wired as a dependency via a group. Any cell
which provides a ConfigReconciler with the correct group name will be
added to the list. Since slice order isn't guaranteed, a priority system
is added.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bgp: Add DiffStore

This commits adds a `DiffStore`. A DiffStore is created from a resource
and embeds its `resource.Store`, thus providing full store capabilities.
Additionally the `DiffStore` records which keys have been changed since
the last time `Diff` was called. `Diff` returns a list of upserted
objects and deleted keys.

This feature will allow us to implement partial updates for
service announcements within the level triggered reconciliation
architecture of the BGP Control plane. This is needed since the amount
of service resources is significantly higher(>750k in some use cases)
than the amount of BGPPeeringPolicy objects.

The `DiffStore` will also trigger the `Signaler` when changes are
detected and signal the controller to reconcile.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
docs: Add documentation for BGP-CP loadbalancer service announcements

This commit updates the existing BGP Control Plane docs to explain the
new field in the CRD and how to use it.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joamaki commented on pkg/k8s/node.go:
```diff
@@ -214,5 +214,8 @@ func ParseNode(k8sNode *slim_corev1.Node, source source.Source) *nodeTypes.Node
 		}
 	}
 
+	newNode.Labels = k8sNode.GetLabels()
```
`Labels` is already assigned on line 115. Move `newNode.Annotations` next to it?

joamaki commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -5,100 +5,108 @@ package agent
+
+	select {
+	case <-s.doneChan:
+	case <-s.ctx.Done():
```
missing `return s.ctx.Err()`

joamaki commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -5,100 +5,108 @@ package agent
-}
+func (s *localNodeStoreSpecer) Start(ctx hive.HookContext) error {
+	s.LocalNodeStore.Observe(s.ctx, func(_ types.Node) {
+		s.Signaler.Event(nil)
```
Instead of using `LocalNodeStore.Get()` down below, considering maintaining a local copy instead and have getters access that. Feels more consistent and less copying of the full `types.Node`.

Or better yet, drop the whole `nodespecer.go` thing, do `stream.Map` from types.Node to the few fields needed and subscribe to that instead. Ok with punting that to later PR to avoid changing the structure too much yet.

dylandreimerink replied:
Yes, agreed. Ideally, we can do without nodespecer and give every cell direct access to the local node store or a resource for the local node. However, this would have required me to touch the existing reconciles which I am not willing to do in this PR. Will tackle this at some point in the future.

joamaki commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -191,8 +183,24 @@ func NewController(params ControllerParams) (*Controller, error) {
 // Start is called by hive after all of our dependencies have been started.
-func (c *Controller) Start(_ hive.HookContext) error {
-	go c.policyInformer.Run(c.ctx.Done())
+func (c *Controller) Start(ctx hive.HookContext) error {
```
Might be worth naming the hook context `startCtx` when there's multiple contexts used.

joamaki commented on pkg/bgpv1/cell.go:
```diff
@@ -19,7 +25,21 @@ var Cell = cell.Module(
+func newBGPPeeringPolicyResource(lc hive.Lifecycle, c client.Clientset) resource.Resource[*v2alpha1api.CiliumBGPPeeringPolicy] {
+	var listerWatcher v2alpha1.CiliumBGPPeeringPolicyInterface
+	if c.IsEnabled() {
+		listerWatcher = c.CiliumV2alpha1().CiliumBGPPeeringPolicies()
```
This should return `nil` resource instead as using a nil lister watcher probably punts the failure to when something tries to use the resource. Likely much easier to figure out what's going on when the resource itself is nil deref'd.

joamaki commented on pkg/bgpv1/gobgp/diffstore.go:
```diff
@@ -0,0 +1,165 @@
+
+	select {
+	case <-ds.doneChan:
+	case <-stopCtx.Done():
```
missing `return stopCtx.Err()`.

joamaki commented on pkg/bgpv1/gobgp/diffstore_mock.go:
```diff
@@ -0,0 +1,129 @@
+
+var _ DiffStore[*runtime.Unknown] = (*mockDiffStore[*runtime.Unknown])(nil)
+
+type mockDiffStore[T runtime.Object] struct {
```
nit: this is a fake not a mock. https://testing.googleblog.com/2013/07/testing-on-toilet-know-your-test-doubles.html.

dylandreimerink replied:
Today I learned

joamaki commented on pkg/bgpv1/cell.go:
```diff
@@ -45,3 +57,25 @@ func newBGPPeeringPolicyResource(lc hive.Lifecycle, c client.Clientset) resource
 }
+
+// Constructs a slim service resource with perpetual retries at a exponential backoff
+func newSlimServiceService(lc hive.Lifecycle, c k8sClient.Clientset) (resource.Resource[*slim_core_v1.Service], error) {
```
s/newSlimServiceService/newSlimServiceResource/

joamaki commented on pkg/bgpv1/cell.go:
```diff
@@ -32,6 +40,10 @@ var Cell = cell.Module(
+	// Create a slim service resource
+	cell.ProvidePrivate(newSlimServiceService),
+	// Create a slim service DiffStore
+	cell.ProvidePrivate(gobgp.NewDiffStore[*slim_core_v1.Service]),
```
consider collapsing the providers:
```
cell.ProvidePrivate(
  newSlimServiceResource,
  gobgp.NewDiffStore[*slim_core_v1.Service],
)
```

I would leave out comments like "Create a slim service resource" as it adds no new information.

joamaki commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -460,3 +467,256 @@ func exportPodCIDRReconciler(ctx context.Context, _ *BGPRouterManager, sc *Serve
+
+	var desiredRoutes []*net.IPNet
+	for _, ingress := range svc.Status.LoadBalancer.Ingress {
+		// Ignore ingresses without ips
```
nit: don't add comments that repeat the code. It should be obvious to the reader that `if ingress.IP == "" { continue }` means "ignore ingresses without IPs".

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -49,6 +54,9 @@ var (
 		// daemonCell wraps the legacy daemon initialization and provides Promise[*Daemon].
 		daemonCell,
+
+		// The BGP Control Plane
```
I would expand this comment to have few more details (max 1-2 lines) and then document in bgpv1.Cell 
in more detail what it does (e.g. listens for these and these k8s events and talks to BGP in this way).

joamaki commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -221,33 +250,45 @@ func NewController(ctx context.Context, clientset client.Clientset, rtMgr BGPRou
+
+	select {
+	case <-ctx.Done():
+	case <-c.doneChan:
```
What about waiting for `policyInformer.Run` and `NodeSpec.Run`? Consider using workerpool in follow-up PR.

joamaki commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -221,33 +250,45 @@ func NewController(ctx context.Context, clientset client.Clientset, rtMgr BGPRou
+	c.cancel()
+
+	select {
+	case <-ctx.Done():
```
oops missed on first round, `return ctx.Err()`.

joamaki commented on pkg/bgpv1/gobgp/reconcile_test.go:
```diff
@@ -470,3 +473,516 @@ func TestExportPodCIDRReconciler(t *testing.T) {
+						}
+					}
+					if !seen {
+						t.Fatalf("unwated advert %+v", advrt)
```
s/unwated/unwanted/

joamaki commented on pkg/bgpv1/gobgp/reconcile_test.go:
```diff
@@ -12,6 +12,9 @@ import (
 	"github.com/cilium/cilium/pkg/bgpv1/agent"
 	v2alpha1api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
```
the convention we've been using for these imports is `slim_corev1` and `slim_metav1`.

joamaki commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -460,3 +466,255 @@ func exportPodCIDRReconciler(ctx context.Context, _ *BGPRouterManager, sc *Serve
+// fullReconciliation reconciles all services, this is a heavy operation due to the potential amount of services and
+// thus should be avoided if partial reconciliation is an option.
+func (r *LBServiceReconciler) fullReconciliation(ctx context.Context, m *BGPRouterManager, sc *ServerWithConfig, newc *v2alpha1api.CiliumBGPVirtualRouter, cstate *agent.ControlPlaneState) error {
+	log.Debug("Start full SVC-LB reconciliation")
```
Informationless debug logging isn't very useful. Consider removing all debug logging and do metrics instead. If you think debug logging is essential, then log at the end with summary of what happened.

joamaki replied:
Longer-term I think we'll want centralized status gathering where we could report stuff like when certain operations ran, how long it took and whether it failed. That's likely much better than metrics as it can carry more information. We can then derive metrics from that. For now though debug logs are fine, as long as they're useful enough.

dylandreimerink replied:
At some point I used this statement to check if the logic worked, forgot to remove it, nice catch

joamaki commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -460,3 +466,255 @@ func exportPodCIDRReconciler(ctx context.Context, _ *BGPRouterManager, sc *Serve
+// svcDiffReconciliation performs reconciliation, only on services which have been created, updated or deleted since
+// the last diff reconciliation.
+func (r *LBServiceReconciler) svcDiffReconciliation(ctx context.Context, m *BGPRouterManager, sc *ServerWithConfig, newc *v2alpha1api.CiliumBGPVirtualRouter, cstate *agent.ControlPlaneState) error {
+	log.Debug("Start diff SVC-LB reconciliation")
```
ditto here about debug logging.

joamaki commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -460,3 +466,255 @@ func exportPodCIDRReconciler(ctx context.Context, _ *BGPRouterManager, sc *Serve
+	}
+
+	if svc.Spec.ExternalTrafficPolicy == slimv1.ServiceExternalTrafficPolicyTypeLocal {
+		// TODO check if "we" the local node have at least 1 endpoint for this service. If not don't add
```
unaddressed TODO

joamaki commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -10,10 +10,16 @@ import (
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slimmetav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
```
s/slimv1/slim_corev1/, s/slimmetav1/slim_metav1/ so it'll follow conventions used everywhere else.

ldelossa commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -5,100 +5,120 @@ package agent
-// The BGP Control Plane needs to obtain Node information however, it may need
-// to query the Kubernetes Node or the Cilium Node object depending on IPAM
-// configuration.
 type nodeSpecer interface {
```
I think we still want some type of comments here, and in this file, no?

ldelossa commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -5,100 +5,120 @@ package agent
-	n, err := s.NodeLister.Get(s.Name)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get node %v: %v", s.Name, err)
+func NewLocalNodeStoreSpecer(params localNodeStoreSpecerParams) (nodeSpecer, error) {
```
Comments

ldelossa commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -5,100 +5,120 @@ package agent
+func (s *localNodeStoreSpecer) Start(_ hive.HookContext) error {
+	s.workerpool.Submit("local-node-store-observe", func(ctx context.Context) error {
+		doneChan := make(chan struct{})
+		s.LocalNodeStore.Observe(ctx, func(_ types.Node) {
```
just a nit, but this would be far easier to read if either the functions werent inline or were formatted each on new lines.

ldelossa commented on pkg/bgpv1/agent/nodespecer.go:
```diff
@@ -5,100 +5,120 @@ package agent
+
+func (s *localNodeStoreSpecer) Annotations() (map[string]string, error) {
+	n := s.LocalNodeStore.Get()
 	return n.Annotations, nil
```
nit, just get rid of the "n" tmp var?

ldelossa commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -95,17 +91,21 @@ type ControlPlaneState struct {
 }
 
+type policyLister interface {
+	List() []*v2alpha1api.CiliumBGPPeeringPolicy
```
Why avoid the possibility to return an error? Multiple implementations could exist, would it be possible that one might want to return an error at some point?

ldelossa commented on pkg/bgpv1/cell.go:
```diff
@@ -19,7 +24,20 @@ var Cell = cell.Module(
 	cell.ProvidePrivate(gobgp.NewBGPRouterManager),
 )
+
+func newBGPPeeringPolicyResource(lc hive.Lifecycle, c client.Clientset) resource.Resource[*v2alpha1api.CiliumBGPPeeringPolicy] {
```
Is the hive passing down a client.Clientset to this?

dylandreimerink replied:
Yes, these parameters are wired via hive.

ldelossa commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -12,57 +12,48 @@ import (
+	NewExportPodCIDRReconciler,
+)
+
+type preflightReconcilerOut struct {
```
What is the major advantage to putting business logic like this into the dependency graph. Wouldn't it be more appropriate to have the controlling structure which *calls* this business logic be in the DI graph, and not have  every bit of logic that goes into the cell?

dylandreimerink replied:
The motivation here is that different reconcilers need different data. Adding all dependencies via the function parameters doesn't seem like a good way to go to me.

Every time an additional dependency is needed for a new or existing reconcilers, the function signature has to change, which breaks all other reconcilers. Ideally we have a very stable interface for reconcilers, and they can get their dependencies via DI.

ldelossa replied:
I see, and since the controlling structure may not know all reconcilers which are instantiated, it can't create the depedencies for each one at it's level. 

Ok I think this makes sense. I do have some concern about how far this goes, and how deep we keep extended the DI graph, but maybe its not a big deal.

ldelossa commented on pkg/bgpv1/gobgp/routermgr.go:
```diff
@@ -36,6 +38,12 @@ var (
 // configuration info.
 type LocalASNMap map[int]*ServerWithConfig
 
+type bgpRouterManagerIn struct {
```
The naming here seem to be inconsistent with the Controller's params.

```
// ControllerParams contains all parameters needed to construct a Controller
type ControllerParams struct {
	cell.In

	Lifecycle    hive.Lifecycle
	Clientset    client.Clientset
	Sig          Signaler
	RouteMgr     BGPRouterManager
	DaemonConfig *option.DaemonConfig
	NodeSpec     nodeSpecer
}
```

Is that intentional?

dylandreimerink replied:
Nope.

ldelossa commented on pkg/bgpv1/gobgp/routermgr.go:
```diff
@@ -296,11 +310,43 @@ func (m *BGPRouterManager) reconcile(ctx context.Context, rd *reconcileDiff) err
 	return nil
 }
+
+// reconcileBGPConfig will utilize the current set of ConfigReconcilerFunc(s)
```
Is "ConfigReconcilerFunc(s)" still the right terminology for this function after the move to the interfaces?

ldelossa commented on pkg/bgpv1/gobgp/diffstore.go:
```diff
@@ -0,0 +1,166 @@
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+)
+
+// DiffStore is a super set of the resource.Store. The diffStore tracks all changes made to the diff store since the
```
```suggestion
// DiffStore is a super set of the resource.Store. The diffStore tracks all changes made to it since the
```

ldelossa commented on pkg/bgpv1/gobgp/diffstore.go:
```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
```
is anything about this DiffStore `gobgp` specific?

maybe consider making this a higher level page and not nested into the `gobgp` namespace.

dylandreimerink replied:
Yes. I did consider this. Even had a PR out for it #22212. There are two reasons for having it in the gobgp namespace. 

First is that we use this custom `Signaler` instead of a generic mechanism, so it is just easier for now.

Second is that this DiffStore shouldn't be the long term solution, not something we want to promote if not necessary. It is a temporary solution to do differential updates within the current architecture, but long term we want to move to using `Resource[T]` directly, which will improve memory overhead and add retry logic.

ldelossa replied:
Gotchya.

ldelossa commented on pkg/bgpv1/gobgp/diffstore.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+func (ds *diffStore[T]) handleEvent(event resource.Event[T]) {
+	update := func(k resource.Key, t T) error {
```
```suggestion
        update := func(k resource.Key, _ T) error {
```

ldelossa commented on pkg/bgpv1/gobgp/diffstore.go:
```diff
@@ -0,0 +1,166 @@
+	return nil
+}
+
+func (ds *diffStore[T]) run() {
```
Will the diff store handle events in the same exact order as the Resource generates them? 

There maybe some out of order concerns here if there are any thread parks or intermediate pauses between a resource's events and the handling of it. 

It appears all events are being forwarded to updateChan, but is the event source from there consistently ordered with the resource's events?

dylandreimerink replied:
I can't imagine, given the level-trigger nature of k8s resource handling that the order of updates should matter. If it does, we should update the diffstore since the `updatedKeys` map will snuffle everything.

ldelossa commented on pkg/bgpv1/gobgp/diffstore.go:
```diff
@@ -0,0 +1,166 @@
+	}
+
+	// Deleting keys doesn't shrink the memory size, so re-create the map to reduce memory usage
+	sd.updatedKeys = make(map[resource.Key]bool, 0)
```
just food for thought, 

thrashing Diff() may make this realloc not that great. 

Maybe it makes sense to only realloc if the slice got larger then a certain reasonable size?

ldelossa commented on daemon/cmd/cells.go:
```diff
@@ -55,7 +55,8 @@ var (
 		daemonCell,
 
-		// The BGP Control Plane
+		// The BGP Control Plane, when enabled, can announce routes via BGP so that traffic for those
```
It's a bit hard to summarize all the things the BGP-CP may do, and also keep it up to date. For instance, when we begin learning routes, this is now an outdated statement. 

It maybe better to keep this vague? I.e. "The BGP Control Plane which enables various BGP related interop"

ldelossa commented on pkg/bgpv1/cell.go:
```diff
@@ -43,3 +57,25 @@ func newBGPPeeringPolicyResource(lc hive.Lifecycle, c client.Clientset) resource
+	), nil
+}
+
+func errorRateLimiter() workqueue.RateLimiter {
```
This seems to be a constructor, why not prefix the function with "New"

ldelossa commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -12,57 +12,48 @@ import (
+type ConfigReconciler interface {
+	// Priority is used to determine the order in which reconcilers are called. Reconcilers are called from lowest to
+	// highest.
+	Priority() int
```
Having each ConfigReconciler hardcode its priority seems like a code smell. 

For instance, what happens if someone wants to introduce an out-of-tree or plugable ConfigReconciler, and wants to place its execution somewhere between existing ones. 

There would be no way to do this, previous, at least the slice of ConfigReconcilerFuncs could be reallocated or reordered, but now each Reconciler is defining its order concretely. Is this desired?

dylandreimerink replied:
> For instance, what happens if someone wants to introduce an out-of-tree or plugable ConfigReconciler, and wants to place its execution somewhere between existing ones.

My current answer to that issue is that all existing reconcilers have a spacing of 10, so 9 other reconcilers should be able to order between each existing one.

> There would be no way to do this, previous, at least the slice of ConfigReconcilerFuncs could be reallocated or reordered, but now each Reconciler is defining its order concretely. Is this desired?

Come to think of this, the whole ordering is here since some reconcilers build atop config make by others. But it doesn't matter if the PodCIDR or SvcLB reconciler runs first for example. Perhaps each reconciler should be able to say that is should run after one or more other reconcilers. A bit like the SystemD approach. Then we can sort based on those dependencies without predefined priority and its associated downfalls.

ldelossa replied:
Yeah I do think moving forward we probably don't want to hard code the priority numbers. 

Keeping the gaps in the space is probably OK for now.

ldelossa commented on pkg/bgpv1/gobgp/reconcile.go:
```diff
@@ -460,3 +466,249 @@ func exportPodCIDRReconciler(ctx context.Context, _ *BGPRouterManager, sc *Serve
+
+	if changed {
+		if err := r.fullReconciliation(ctx, m, sc, newc, cstate); err != nil {
+			return fmt.Errorf("Full reconciliation: %w", err)
```
linter is barking that errors start with caps, you'll want to do a pass on errors and lower case their first words.

ldelossa reviewed (changes requested):
Logic looks good. Mostly questions around refactor.

dylandreimerink committed:
bgp,k8s: Added .Spec.serviceSelector to CiliumBGPPeeringPolicy CRD

Added a optional service selector for the CiliumBGPPeeringPolicy. If
set, all services matching the selector will be announced. If the
selector is null/unset we will not announce any services to maintain
consistent behavior for users whom already have peering policies.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
bgp: Added LoadBalancer Service reconciler

This commit adds a new reconciler which manages announcements for
loadbalancer services. This reconcilers utilizes a `DiffStore` to
get a change set since the last time the reconciler was invoked. This
partial synchronization is necessary to be able to handle large
amounts of services.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joestringer reviewed (approved):
I looked at the docs, as well as `pkg/k8s/apis`, and the nodespecer commit that triggers most of the vendor changes. These all LGTM. The only API impact I see is adding a brand new field, but we bump the `CustomResourceDefinitionSchemaVersion` anyway so the newer versions of the CRD should be properly installed into the cluster.
